src/esp/assets/PTexMeshData.cpp

```diff
@@ -220,12 +247,116 @@ std::vector<PTexMeshData::MeshData> splitMesh(
       uint32_t index = refdVerts[j];
       subMeshes[i].vbo[j] = mesh.vbo[index];
       subMeshes[i].nbo[j] = mesh.nbo[index];
+      // Careful:
+      // for Ptex mesh we never ever set the "cbo"
     }
   }
 
   return subMeshes;
 }
 
+// =========== the input file format =======================
+// a uint64_t, N, the number of sub-meshes;
+
+// it follows by N chunks, each of which contains:
+// a uint64_t, M, the number of faces within the chunk;
+// M uint32_t, face indices (sorted) in the *original* mesh;
+// =========================================================
+
+//  this binary can be generated by modifying the splitMesh() in file
+//  PTexLib.cpp in ReplicaSDK.
+//  and the name is hard-coded as "sorted_faces.bin" in the simulator.
+
+// Put it in the sub-folder, "habitat".
+
+std::vector<PTexMeshData::MeshData> loadSubMeshes(
+    const PTexMeshData::MeshData& mesh,
+    const std::string& filename) {
+  // sanity checks
+  CORRADE_ASSERT(!filename.empty(), "Error: filename cannot be empty.", {});
+  std::ifstream file;
+  file.open(filename, std::ios::in | std::ios::binary);
+  CORRADE_ASSERT(file.good(), "Error: cannot open the file " << filename, {});
+
+  uint64_t numSubMeshes = 0;
+  file.read((char*)&numSubMeshes, sizeof(uint64_t));
+
+  std::vector<PTexMeshData::MeshData> subMeshes(numSubMeshes);
+
+  size_t totalFaces = 0;  // used in sanity check
+  for (uint64_t iMesh = 0; iMesh < numSubMeshes; ++iMesh) {
+    uint64_t numFaces = 0;
+    file.read((char*)&numFaces, sizeof(uint64_t));
+
+    std::vector<uint32_t> originalFaces(numFaces);
+    // load the face indices in the *original* mesh
+    file.read((char*)originalFaces.data(), sizeof(uint32_t) * numFaces);
+    // a *vertex* lookup table:
+    // global index of the original mesh --> local index in sub-meshes
+    // (note: this table cannot be defined outside of the for loop, as a vertex
+    // in original mesh may appear in different sub-meshes.)
+    std::unordered_map<uint32_t, uint32_t> globalToLocal;
+
+    // Another *vertex* lookup table:
+    // local index of current sub-mesh --> global index of the original mesh
+    std::vector<uint32_t> localToGlobal;
+
+    // compute the two lookup tables
+    for (size_t jFace = 0; jFace < numFaces; ++jFace) {
+      uint32_t f = originalFaces[jFace];  // face index in original mesh
+      for (size_t v = 0; v < 4; ++v) {
+        uint32_t global = mesh.ibo[f * 4 + v];
+        if (globalToLocal.find(global) == globalToLocal.end()) {
+          globalToLocal[global] = localToGlobal.size();
+          localToGlobal.push_back(global);
+        }
+      }
+    }
+
+    // compute the ibo for the current sub-mesh
+    auto& ibo = subMeshes[iMesh].ibo;
+    ibo.resize(numFaces * 4);
+    uint64_t idx = 0;
+    for (size_t jFace = 0; jFace < numFaces; ++jFace) {
```

Comment:
Couldn't this be done in the loop above. iow, could we collapse these two loops.

Reply:
Certainly you can. 
But it would reduce the code readability, and thus cause difficulty to the future maintenance in my humble opinion. 
Unless the the jobs are extremely simple and trivial, I  prefer just to do "one job" in one for loop.

Here, as the comments suggest, the steps are quite clear by having individual loops:
- compute the two lookup tables, localToGlobal, globalToLocal;
- compute ibo based on globalToLocal;
- compute vbo based on localToGlobal;

Reply:
Agree. Readability is important unless it affects performance critical. A single loop would be more cache efficient and we are processing a lot of data. If this code path taken some time you may want to consider a single loop.

Reply:
> Agree. Readability is important unless it affects performance critical. A single loop would be more cache efficient and we are processing a lot of data. If this code path is taking some time you may want to consider a single loop. Maybe time this loop on a large mesh?

This function is called only once in the initial stage and the time complexity is O(n). The largest model contains roughly 9M triangles. So it is totally fine.